README.md
# Anime Generator Stable Diffusion Docker Container

This Docker container is a text-to-image (txt2img) generation tool that leverages Stable Diffusion with a packaged diffusion model. Stable Diffusion is a deep learning, latent txt2img diffusion model capable of generating photo-realistic images given any text input, for more information visit this wikipedia page: [Stable Diffusion](https://en.wikipedia.org/wiki/Stable_Diffusion). The diffusion model that was packaged in this docker container can be found here: [Diffuser Model](https://huggingface.co/ckpt/anything-v4.5-vae-swapped). This Docker container is built with horizontal scaling in mind, and as such will be able to placed on cloud provider services like AWS, GCP, and Azure.

## Example Image Output

These are some of the example images with their corresponding commands. These commands are used to generate them using artificial intelligence. Feel free to test this out yourself!

```sh
./build.sh run --prompt "highres, 1girl, purple hair, long hair, yellow eyes, black warrior armor, warrior princess, battlefield, shadows, lens flare, masterpiece, sunshine, clothed, red lips, forest, breathtakingly beautiful, hair bun, tanned skin" --negative-prompt "lowres, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry, artist name, NSFW, cat ears, weapon, sword, shield"
```

### Purple-Haired Warrior Princess

<p align="center">
  <img src="img/highres,_1girl,_purple_hair,_long_hair,_yellow_eyes,_black_warrior_armor,_warrior_princess,_battlefield,_shadows,_lens_flare,_masterpiece,_sunshine,_clothed,_red_lips,_fo__steps_22__scale_11.00__seed_8571956922351865579__n_1.png" alt="Purple-Haired Warrior Princess" width="256" height="256">
  <img src="img/highres,_1girl,_purple_hair,_long_hair,_yellow_eyes,_black_warrior_armor,_warrior_princess,_battlefield,_shadows,_lens_flare,_masterpiece,_sunshine,_clothed,_red_lips,_fo__steps_22__scale_11.00__seed_11254028509881136190__n_1.png" alt="Purple-Haired Warrior Princess" width="256" height="256">
  <img src="img/highres,_1girl,_purple_hair,_long_hair,_yellow_eyes,_black_warrior_armor,_warrior_princess,_battlefield,_shadows,_lens_flare,_masterpiece,_sunshine,_clothed,_red_lips,_fo__steps_22__scale_11.00__seed_14222826273838498285__n_1.png" alt="Purple-Haired Warrior Princess" width="256" height="256">
</p>

These three images are my rendition on a warrior with purple hair, yellow eyes on a battlefield. You can render these same exact images by copying the seed found in the image's names (more information on using seeds below).

### Beautiful Fantasy Village

```sh
./build.sh run --W 960 --H 320 --prompt "beautiful, fantasy, lens flair, masterpiece, sunny, bright, vibrant, cloudy, worried, neon, sparkling, radiant, sunshine, starry, dreams, background, village, grass, town, temple, emissive, utopia, castle" --negative-prompt "lowres, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry, artist name, boy, girl, people, face" --ddim_steps 25
```

<p align="center">
  <img src="img/beautiful,_fantasy,_lens_flair,_masterpiece,_sunny,_bright,_vibrant,_cloudy,_worried,_neon,_sparkling,_radiant,_sunshine,_starry,_dreams,_background,_village,_grass,_town__steps_25__scale_11.00__seed_3463221830618119851__n_1.png" alt="Beautiful Fantasy Village" width="768" height="256">
</p>

This image is a rendition of a beautiful village with a bright sky. You can render this same exact image by copying the seed found in the image's name (more information on using seeds below).

## Requirements

By default, this pipeline will render images using CPU by default. This is intended as rendering images on a GPU is extremely cost prohibitive. As a result of this, it will take a few minutes to create one image (Roughly 5 minutes and 20 seconds on a Ryzen 5 5600x CPU using the ```test``` option down below, your mileage may vary). The bundled dockerfile runs on 3.10-slim-bullseye, and as such is tested on **Debian 11** based systems. The operating System I used to personally test this docker container was "Ubuntu Jammy 22.04 (LTS)". If you're using **Alpine**, or **Windows**, or **macOS** it may not work, and as such, your mileage may vary (Will check compatability at a later date).

## Quickstart (These commands are for the following OS: "Ubuntu Jammy 22.04 (LTS)")

### Docker setup (Ignore if you already have docker installed on your OS)

It is important to have docker engine installed prior to running this project. If you don't have it installed, follow the official [docker setup](https://docs.docker.com/engine/install/ubuntu/) guide.

### Git and Latest-version setup

The pipeline is managed using a single [`build.sh`](build.sh) script. Big kudos to [fboulnois](https://github.com/fboulnois) for setting this up. I cannot thank you enough. 

Start with the following command to pull from repo:

```sh
git clone https://github.com/azher-alnabi/anime-stable-diffusion-docker-container.git && cd anime-stable-diffusion-docker-container
```

Then run the following command to setup the latest version of this repo:

```sh
./build.sh pull
```

## Build

First begin by building with the following command:

```sh
./build.sh build
```

## Testing

Run a test using the following command to see if an image has generated. Navigate to the `output` folder to view this image:

```sh
./build.sh test
```

This test should render a single 512x512 (height and width) image similar to the three images shown above.

## Run

### Text-to-Image 

Now run whatever prompts you want with the following command. Try out the options down below (Remember, by default it uses CPU to render images):

```sh
./build.sh run --prompt 'warrior princess'
```

### Options

Some of the options from [`txt2img.py`](https://github.com/CompVis/stable-diffusion/blob/main/scripts/txt2img.py) are implemented for compatibility:

* `--prompt [PROMPT]`: the prompt to render into an image
* `--negative-prompt [NEGATIVE_PROMPT]`: the prompt to not render into an image (default `None`)
* `--n_samples [N_SAMPLES]`: number of images to create per run (default 1)
* `--n_iter [N_ITER]`: number of times to run pipeline (default 1)
* `--H [H]`: image height in pixels (default 512, must be divisible by 64)
* `--W [W]`: image width in pixels (default 512, must be divisible by 64)
* `--scale [SCALE]`: Classifier Guidance Scale (CGS) (default 11)
* `--seed [SEED]`: RNG seed for repeatability (default is a random seed)
* `--ddim_steps [DDIM_STEPS]`: number of sampling steps (default 22)
* `--skip`: skip safety checker (default is the safety checker is on)
* `--model [MODEL]`: the model used to render images (default is `ckpt/anything-v4.5-vae-swapped`)
* `--scheduler [SCHEDULER]`: override the scheduler used to denoise the image (default `DPMSolverSinglestepScheduler`)

## Outputs

### Images

The images are saved as PNGs in the `output` folder using the prompt text. The `build.sh` script creates and mounts this folder as a volume in the docker container.

## Examples

These commands are both identical:

```sh
./build.sh run 'warrior princess'
./build.sh run --prompt 'warrior princess'
```

This command enables GPU rendering (Must use a Nvidia GPU) using the following command:

```sh
./build.sh run --device gpu --prompt 'warrior princess'
```

This command sets the seed to 255 (Allows rendering of image with constant output):

```sh
./build.sh run --seed 255 --prompt 'warrior princess'
```

Several commands can be combined together:

```sh
./build.sh run --skip --scale 11.0 --seed 255 --prompt 'warrior princess'
```

## Minimize Time Rendering

There are a few steps you can include if you want to minimize the time it takes to render an image.

* Make images smaller than 512x512 using `--W` and `--H` to decrease memory use and increase image creation speed
* Decrease the number of samples and increase the number of iterations with `--n_samples` and `--n_iter` to decrease overall memory use
* Skip the safety checker with `--skip` to run less code (Work in progress: Implementation by default)

### CPU Time Minimization

```sh
./build.sh run --skip \
  --W 256 --H 256 \
  --n_samples 1 --n_iter 1 --prompt 'warrior princess'
```

### GPU Time Minimization (Experimental, please open up an issue and let me know if any features are not working)

```sh
./build.sh run --device cuda --skip \
  --W 256 --H 256 --half --attention-slicing \
  --n_samples 1 --n_iter 1 --prompt 'warrior princess'
```

* `--attention-slicing`: use less memory at the expense of inference speed (default is no attention slicing)
* `--device [DEVICE]`: the cpu or cuda device to use to render images (default is `cpu`, switch to `cuda` for GPU rendering)
* `--half`: use float16 tensors instead of float32 (default `float32`)

## Running on Windows (Experimental, please open up an issue and let me know if any features are not working)

On Windows, if you aren't using WSL2 and instead use MSYS, MinGW, or Git Bash,
prefix your commands with `MSYS_NO_PATHCONV=1` (or export it beforehand):

```sh
MSYS_NO_PATHCONV=1 ./build.sh run --prompt 'warrior princess'
```

## Token setup (Experimental, please open up an issue and let me know if any features are not working)

As far as I am aware, you don't need to pass through a huggingface token currently, however open a github issue if an error occurs. The solution to a token problem may involve saving the [user access token](#huggingface-token) in a file called `token.txt`. Another way to solve this problem would be by adding the token manually using the following command:

* `--token [TOKEN]`: specify a Huggingface user access token at the command line instead of reading it from a file (default is a file)

## Docker Cleanup and Pruning

If you want to remove all docker images and containers, you can use the following command:

```sh
./build.sh clean
```

If you want to remove all docker images, containers, volumes, and output images, you can use the following command:

```sh
./build.sh wipe
```

## Disclaimer

By default, the default model that is being used in this docker image will render images that are **SFW** **(_Safe For Work_)**. As a result of this, the outputted images may be returned blank if an image is flagged for being **NSFW** **(_Not Safe For Work_)**. The built-in saftey checker is responsible for this, and may cause problems due to it being. If you want to skip over this saftey feature, and render images that have the potential for being **NSFW**, you can use the ```--skip``` feature set to disable the saftey checker. **Use at your own discretion**.

## Credits

Special Thanks to the following people and communities involved with helping me create this Docker container.

### Github Member: fboutnois
Thank you to fboutnois for the [original](https://github.com/fboulnois/stable-diffusion-docker) implementation of this Docker container, and for being super accommodating to my questions and requests. Check out their [github](https://github.com/fboulnois) profile for more of their work.

### HuggingFace Member: Furqanil Taqwa (Aka Linaqruf)
Thank you to Linaqruf for the creation of the base diffuser model: Linaqruf/anything-v3-better-vae. Their work can be found at: [Linaqruf Hugging Face Profile](https://huggingface.co/Linaqruf)

### HuggingFace and Github Member: Camenduru
Thank you to Camenduru for archiving the diffusers that was used in this docker container. Check out their [github](https://github.com/camenduru) and their [Hugging Face](https://huggingface.co/ckpt) profiles.

### Huggingface Community
Thank you to the Huggingface team for creating the AI community and Machine Learning platform, community found at: [Hugging Face Community Landing Page](https://huggingface.co/).

### Stability AI
Thank you to Stability AI for open sourcing Stable Diffusion. Learn more about their work at: [Stability AI Landing Page](https://stability.ai/).

CHANGELOG.md
# Changelog

## [v1.1.0](https://github.com/azher-alnabi/anime-stable-diffusion-docker-container/compare/v1.0.0...v1.1.0) - 2023-02-03

### Replaced

* Default Diffuser: Anything-v3-vae -> Anything-v4.5-vae
* Default step size: 20 -> 22

### Added 

* Docker wipe script: ./build.sh wipe

## [v1.0.0](https://github.com/azher-alnabi/anime-stable-diffusion-docker-container/compare/v1.0.0-alpha...v1.0.1-alpha) - 2023-01-31

### Added

* Anime Model

### Removed

* Img2Img
* Inpainting
* 4x Upscaling

## [v1.27.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.26.0...v1.27.0) - 2023-01-16

### Added

* Automatically publish new versions

### Fixed

* Update unstable xformers to 0.0.16rc425

## [v1.26.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.25.1...v1.26.0) - 2023-01-13

### Added

* Support pulling image from ghcr
* Publish docker pipeline to ghcr

### Changed

* Simplify ghcr publish action
* Move ghcr url into variable

## [v1.25.1](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.25.0...v1.25.1) - 2023-01-10

### Fixed

* Update safetensors and unstable xformers
* Add numpy 1.23.5 to fix float errors

## [v1.25.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.24.0...v1.25.0) - 2022-12-26

### Added

* Add test for scheduler option
* Allow different schedulers to be used
* Update torch to 1.13.1 and unstable xformers

## [v1.24.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.23.0...v1.24.0) - 2022-12-20

### Added

* Add tests for stable diffusion 2.0 and 2.1
* Update diffusers to 0.11.1
* Switch to python slim and halve image size

## [v1.23.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.22.0...v1.23.0) - 2022-12-14

### Added

* Add memory efficient transformers
* Add test for image upscaling

### Fixed

* Ensure default test produces output

## [v1.22.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.21.0...v1.22.0) - 2022-12-12

### Added

* Add upscaler pipeline
* Update diffusers to 0.10.2

## [v1.21.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.20.0...v1.21.0) - 2022-12-09

### Added

* Update diffusers to 0.10.1
* Update diffusers to 0.10.0
* Update transformers to 4.25.1

### Fixed

* Remove use of autocast
* Remove unused arguments from pipeline
* Return after setting gpu arg
* Removed unused imports

## [v1.20.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.19.1...v1.20.0) - 2022-11-29

### Added

* Update diffusers to 0.9.0
* Update tensorflow to 2.11.0

### Fixed

* Ensure skip option works on onnx pipeline

## [v1.19.1](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.19.0...v1.19.1) - 2022-11-22

### Added

* Test image to image pipeline

### Fixed

* Do not pass gpu arg if using cpu pipeline

## [v1.19.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.18.0...v1.19.0) - 2022-11-17

### Added

* Add test for cpu pipeline
* Add cpu pipeline using onnx

### Changed

* Test image with different dimensions

## [v1.18.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.17.0...v1.18.0) - 2022-11-11

### Added

* Add diffusion inpainting

## [v1.17.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.16.0...v1.17.0) - 2022-11-08

### Added

* Add image-to-image diffusion
* Add input folder for image-to-image diffusion

### Changed

* Call run command directly from tests

### Fixed

* Ensure output folder is always created

## [v1.16.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.15.0...v1.16.0) - 2022-11-07

### Added

* Add standard tests for all options
* Upgrade diffusers to 0.7.2

## [v1.15.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.14.0...v1.15.0) - 2022-11-04

### Added

* Add negative prompts

### Changed

* Switch to using num_images_per_prompt

## [v1.14.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.13.0...v1.14.0) - 2022-11-03

### Added

* Upgrade diffusers to 0.7.0

## [v1.13.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.12.0...v1.13.0) - 2022-11-02

### Added

* Upgrade torch to 1.13.0 and cuda to 11.7

## [v1.12.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.11.1...v1.12.0) - 2022-11-01

### Added

* Update transformers to v4.24.0

### Fixed

* Always use torch instead of tensorflow

## [v1.11.1](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.11.0...v1.11.1) - 2022-10-26

### Changed

* Split pipeline and inference

## [v1.11.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.10.0...v1.11.0) - 2022-10-21

### Added

* Make model configurable

## [v1.10.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.9.0...v1.10.0) - 2022-10-19

### Added

* Upgrade to diffusers 0.6.0

### Fixed

* Switch output and image names

## [v1.9.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.8.1...v1.9.0) - 2022-10-14

### Added

* Upgrade to latest huggingface releases

## [v1.8.1](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.8.0...v1.8.1) - 2022-10-13

### Fixed

* Remove cuda stubs to prevent errors

## [v1.8.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.7.0...v1.8.0) - 2022-10-12

### Added

* Upgrade to diffusers 0.4.2

## [v1.7.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.6.0...v1.7.0) - 2022-10-07

### Added

* Upgrade to diffusers 0.4.1
* Pin dependencies using requirements.txt

## [v1.6.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.5.0...v1.6.0) - 2022-09-14

### Added

* Add attention slicing

### Changed

* Use tagged version of tensorflow

## [v1.5.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.4.0...v1.5.0) - 2022-09-08

### Added

* Specify user access token at command line

## [v1.4.0](https://github.com/fboulnois/stable-diffusion-docker/compare/v1.3.0...v1.4.0) - 2022-09-05


build.sh
#!/bin/sh

set -eu

CWD=$(basename "$PWD")

set_gpu_arg() {
    while [ "$#" -gt 0 ]; do
        if [ "$1" = "--device" ] && [ "$2" = "gpu" ]; then
            GPU_ARG="--gpus=all"
            return
        fi
        shift
    done
    GPU_ARG=""
}

build() {
    docker build . --tag "$CWD"
}

clean() {
    docker system prune -a -f
}

wipe() {
    docker system prune -a -f 
    docker volume prune -f
    rm -rf output/*
}

dev() {
    docker run --rm --gpus=all --entrypoint=sh \
        -v huggingface:/home/huggingface/.cache/huggingface \
        -v "$PWD"/output:/home/huggingface/output \
        -it "$CWD"
}

pull() {
    GHCR="ghcr.io/azher-alnabi/anime-stable-diffusion-docker-container"
    docker pull "$GHCR"
    docker tag "$GHCR" "$CWD"
}

run() {
    set_gpu_arg "$@"
    docker run --rm ${GPU_ARG} \
        -v huggingface:/home/huggingface/.cache/huggingface \
        -v "$PWD"/output:/home/huggingface/output \
        "$CWD" "$@"
}

tests() {
    run --skip \
        --prompt "highres, 1girl, purple hair, long hair, yellow eyes, black warrior armor, warrior princess, battlefield, shadows, lens flare, masterpiece, sunshine, clothed, red lips, forest, breathtakingly beautiful, hair bun, tanned skin" \
        --negative-prompt "lowres, bad anatomy, bad hands, text, error, missing fingers, extra digit, fewer digits, cropped, worst quality, low quality, normal quality, jpeg artifacts, signature, watermark, username, blurry, artist name, NSFW, cat ears, weapon, sword, shield" 
}

mkdir -p output
case ${1:-build} in
    build) build ;;
    clean) clean ;;
    wipe) wipe ;;
    dev) dev "$@" ;;
    pull) pull ;;
    run) shift; run "$@" ;;
    test) tests ;;
    *) echo "$0: No command named '$1'" ;;
esac


docker-entrypoint.py
#!/usr/bin/env python
import argparse, datetime, inspect, os
import numpy as np
import torch
from PIL import Image
from diffusers import (
    StableDiffusionPipeline,
    schedulers,
)


def iso_date_time():
    return datetime.datetime.now().isoformat()


def remove_unused_args(p):
    params = inspect.signature(p.pipeline).parameters.keys()
    args = {
        "prompt": p.prompt,
        "negative_prompt": p.negative_prompt,
        "height": p.H,
        "width": p.W,
        "num_images_per_prompt": p.n_samples,
        "num_inference_steps": p.ddim_steps,
        "guidance_scale": p.scale,
        "generator": p.generator,
    }
    return {p: args[p] for p in params if p in args}


def stable_diffusion_pipeline(p):
    p.dtype = torch.float16 if p.half else torch.float32
    p.diffuser = StableDiffusionPipeline
    p.revision = "fp16" if p.half else "main"

    if p.token is None:
        with open("token.txt") as f:
            p.token = f.read().replace("\n", "")

    if p.seed == 0:
        p.seed = torch.random.seed()

    p.generator = torch.Generator(device=p.device).manual_seed(p.seed)

    print("load pipeline start:", iso_date_time(), flush=True)

    pipeline = p.diffuser.from_pretrained(
        p.model,
        torch_dtype=p.dtype,
        revision=p.revision,
        use_auth_token=p.token,
    ).to(p.device)

    if p.scheduler is not None:
        scheduler = getattr(schedulers, p.scheduler)
        pipeline.scheduler = scheduler.from_config(pipeline.scheduler.config)

    if p.skip:
        pipeline.safety_checker = None

    if p.attention_slicing:
        pipeline.enable_attention_slicing()

    p.pipeline = pipeline

    print("loaded models after:", iso_date_time(), flush=True)

    return p


def stable_diffusion_inference(p):
    prefix = p.prompt.replace(" ", "_")[:170]
    for j in range(p.n_iter):
        result = p.pipeline(**remove_unused_args(p))

        for i, img in enumerate(result.images):
            idx = j * p.n_samples + i + 1
            out = f"{prefix}__steps_{p.ddim_steps}__scale_{p.scale:.2f}__seed_{p.seed}__n_{idx}.png"
            img.save(os.path.join("output", out))

    print("completed pipeline:", iso_date_time(), flush=True)


def main():
    parser = argparse.ArgumentParser(description="Create images from a text prompt.")
    parser.add_argument(
        "prompt0",
        metavar="PROMPT",
        type=str,
        nargs="?",
        help="The prompt to render into an image",
    )
    parser.add_argument(
        "--prompt", type=str, nargs="?", help="The prompt to render into an image"
    )
    parser.add_argument(
        "--n_samples",
        type=int,
        nargs="?",
        default=1,
        help="Number of images to create per run",
    )
    parser.add_argument(
        "--n_iter",
        type=int,
        nargs="?",
        default=1,
        help="Number of times to run pipeline",
    )
    parser.add_argument(
        "--H", type=int, nargs="?", default=512, help="Image height in pixels"
    )
    parser.add_argument(
        "--W", type=int, nargs="?", default=512, help="Image width in pixels"
    )
    parser.add_argument(
        "--scale",
        type=float,
        nargs="?",
        default=11,
        help="Classifier free guidance scale",
    )
    parser.add_argument(
        "--seed", type=int, nargs="?", default=0, help="RNG seed for repeatability"
    )
    parser.add_argument(
        "--ddim_steps", type=int, nargs="?", default=22, help="Number of sampling steps"
    )
    parser.add_argument(
        "--attention-slicing",
        type=bool,
        nargs="?",
        const=True,
        default=False,
        help="Use less memory at the expense of inference speed",
    )
    parser.add_argument(
        "--device",
        type=str,
        nargs="?",
        default="cpu",
        help="The cpu or cuda device to use to render images",
    )
    parser.add_argument(
        "--half",
        type=bool,
        nargs="?",
        const=True,
        default=False,
        help="Use float16 (half-sized) tensors instead of float32",
    )
    parser.add_argument(
        "--model",
        type=str,
        nargs="?",
        default="ckpt/anything-v4.5-vae-swapped",
        help="The model used to render images",
    )
    parser.add_argument(
        "--negative-prompt",
        type=str,
        nargs="?",
        help="The prompt to not render into an image",
    )
    parser.add_argument(
        "--scheduler",
        type=str,
        nargs="?",
        default="DPMSolverSinglestepScheduler",
        help="Override the scheduler used to denoise the image",
    )
    parser.add_argument(
        "--skip",
        type=bool,
        nargs="?",
        const=True,
        default=False,
        help="Skip the safety checker",
    )
    parser.add_argument(
        "--token", type=str, nargs="?", help="Huggingface user access token"
    )
    args = parser.parse_args()

    if args.prompt0 is not None:
        args.prompt = args.prompt0
    
    pipeline = stable_diffusion_pipeline(args)
    stable_diffusion_inference(pipeline)


if __name__ == "__main__":
    main()
